myaws/ecs_node_renew.go

```diff
@@ -130,10 +167,60 @@ func (client *Client) ECSNodeRenew(options ECSNodeRenewOptions) error {
 		return err
 	}
 
+	// remove "scale in protection" to instances created at scale-out.
+	fmt.Fprintln(client.stdout, "Removing scale in protection: ", awsutil.Prettify(protectInstanceIds))
+	err = client.AutoScalingSetInstanceProtection(AutoScalingSetInstanceProtectionOptions{
+		options.AsgName,
+		protectInstanceIds,
+		false})
+	if err != nil {
+		return err
+	}
+
 	if err = client.printECSStatus(options.Cluster); err != nil {
 		return err
 	}
 
 	fmt.Fprintln(client.stdout, "end: ecs node renew")
 	return nil
 }
+
+// selectInstanceToProtectFromScaleIn selects instance to protect from Scale in.
+// instance select rule:
+//   instances after scale out - instances before scale out - instances which already set `InstanceProtection==true`
+func (client *Client) selectInstanceToProtectFromScaleIn(oldInstanceIds, allInstanceIds []*string) ([]*string, error) {
+	// get newly created nodes (allInstanceIds - oldInstanceIds)
+	newInstanceIds := difference(allInstanceIds, oldInstanceIds)
+
+	// exclude ProtectedFromScaleIn == true nodes
+	params := &autoscaling.DescribeAutoScalingInstancesInput{
+		InstanceIds: newInstanceIds,
+	}
+	response, err := client.AutoScaling.DescribeAutoScalingInstances(params)
+	if err != nil {
+		return nil, errors.Wrap(err, "DescribeAutoScalingGroups failed:")
+	}
+
+	var targetInstanceIds []*string
+	for _, instance := range response.AutoScalingInstances {
+		if *instance.ProtectedFromScaleIn == false {
+			targetInstanceIds = append(targetInstanceIds, instance.InstanceId)
+		}
+	}
+	return targetInstanceIds, nil
+}
+
+// difference returns the elements in `a` that aren't in `b`.
+func difference(a, b []*string) []*string {
+	mb := make(map[string]struct{}, len(b))
+	for _, x := range b {
+		mb[*x] = struct{}{}
+	}
+	var diff []*string
+	for _, x := range a {
+		if _, found := mb[*x]; !found {
```

Comment:
When checking the existence of a map key, `ok` is used more idiomatically than `found`.

Reply:
fixed.